File: common/src/main/java/app/dropbit/commons/currency/BTCCurrency.java
package app.dropbit.commons.currency;

import android.content.Context;
import android.graphics.drawable.Drawable;
import android.os.Parcel;
import android.os.Parcelable;

import java.math.BigDecimal;

import app.dropbit.commons.R;

public class BTCCurrency extends BaseCurrency implements CryptoCurrency, Parcelable {
    public static final String ALT_CURRENCY_FORMAT = "#,##0.00000000 BTC";
    public static final long MAX_SATOSHI = 2099999997690000L;
    public static final String SYMBOL = "\u20BF";
    public static final Creator<BTCCurrency> CREATOR = new Creator<BTCCurrency>() {
        @Override
        public BTCCurrency createFromParcel(Parcel in) {
            return new BTCCurrency(in);
        }

        @Override
        public BTCCurrency[] newArray(int size) {
            return new BTCCurrency[size];
        }
    };
    static final int SUB_NUM_MAX = 8;
    private static final int WHOLE_NUM_MAX = 8;
    private static final String DEFAULT_CURRENCY_FORMAT = String.format("%s #,##0.########", SYMBOL);
    private static final String INCREMENTAL_FORMAT = NO_SYMBOL_FORMAT;
    private static final String STRING_FORMAT = "#,##0.########";
    private String currencyFormat = DEFAULT_CURRENCY_FORMAT;

    public BTCCurrency() {
        super();
    }

    public BTCCurrency(long initialValue) {
        super(initialValue);
    }

    public BTCCurrency(String initialValue) {
        super(initialValue);
    }

    public BTCCurrency(double initialValue) {
        super(initialValue);
    }

    public BTCCurrency(BigDecimal initialValue) {
        super(initialValue);
    }

    protected BTCCurrency(Parcel in) {
        this(in.readLong());
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeLong(toLong());
    }

    @Override
    public String getSymbol() {
        return SYMBOL;
    }

    @Override
    public String getFormat() {
        return STRING_FORMAT;
    }

    @Override
    public String getCurrencyFormat() {
        return currencyFormat;
    }

    @Override
    public void setCurrencyFormat(String format) {
        currencyFormat = format;
    }

    @Override
    public String getIncrementalFormat() {
        return INCREMENTAL_FORMAT;
    }

    @Override
    public int getMaxNumSubValues() {
        return SUB_NUM_MAX;
    }

    @Override
    public int getMaxNumWholeValues() {
        return WHOLE_NUM_MAX;
    }

    @Override
    public boolean isValid() {
        return (toSatoshis() >= 0 && toSatoshis() <= MAX_SATOSHI);
    }

    @Override
    protected long getMaxLongValue() {
        return MAX_SATOSHI;
    }

    public long toSatoshis() {
        if (isZero()) {
            return 0L;
        }

        BigDecimal myBTCValue = toBigDecimal();
        BigDecimal mySatoshiValue = myBTCValue.multiply(new BigDecimal(100000000));

        return mySatoshiValue.longValue();
    }

    public String toUriFormattedString() {
        BTCCurrency btc = new BTCCurrency(toSatoshis());
        return String.valueOf(btc.value);
    }

    @Override
    public Drawable getSymbolDrawable(Context context) {
        return context.getDrawable(R.drawable.ic_btc_icon);
    }

    @Override
    public FiatCurrency toFiat(FiatCurrency conversionFiat) {
        return toUSD(conversionFiat);
    }
}

File: common/src/main/java/app/dropbit/commons/currency/BTCCurrency.kt
package app.dropbit.commons.currency

import android.content.Context
import android.graphics.drawable.Drawable
import android.os.Parcel
import android.os.Parcelable
import android.os.Parcelable.Creator
import app.dropbit.commons.R.drawable
import java.math.BigDecimal
import java.math.RoundingMode

class BTCCurrency : BaseCurrency, CryptoCurrency, Parcelable {
    override val symbol: String = BTCCurrency.symbol
    override val maxNumSubValues: Int = SUB_NUM_MAX
    override val maxNumWholeValues: Int = 8
    override val format: String = CryptoCurrency.NO_SYMBOL_FORMAT
    override val incrementalFormat: String = CryptoCurrency.NO_SYMBOL_FORMAT
    override var currencyFormat = DEFAULT_CURRENCY_FORMAT

    override val maxLongValue: Long get() = BTCCurrency.maxLongValue

    constructor()
    constructor(initialValue: Long) : this() {
        value = scale(BigDecimal(initialValue).movePointLeft(maxNumSubValues))
    }

    constructor(initialValue: Double) : this() {
        value = BigDecimal(initialValue).stripTrailingZeros().setScale(maxNumSubValues, RoundingMode.HALF_UP)
    }

    constructor(initialValue: BigDecimal) : this() {
        value = initialValue.stripTrailingZeros().setScale(maxNumSubValues, RoundingMode.HALF_UP)
    }

    internal constructor(parcel: Parcel) : this(parcel.readLong())

    override fun describeContents(): Int {
        return 0
    }

    override fun writeToParcel(dest: Parcel, flags: Int) {
        dest.writeLong(toLong())
    }

    override fun isValid(): Boolean {
        return toLong() >= 0 && toLong() <= this.maxLongValue
    }

    fun toUriFormattedString(): String {
        val btc = BTCCurrency(toLong())
        btc.currencyFormat = "#,##0.00000000"
        return btc.toFormattedCurrency()
    }

    override fun toUSD(conversionValue: Currency): USDCurrency {
        if (conversionValue.isZero) {
            return USDCurrency()
        }
        return USDCurrency(toBigDecimal().multiply(conversionValue.toBigDecimal()))
    }

    override fun toSats(conversionValue: Currency): SatoshiCurrency {
        return SatoshiCurrency(toLong())
    }

    override fun toBTC(conversionValue: Currency): BTCCurrency {
        return this
    }

    override fun getSymbolDrawable(context: Context): Drawable? {
        return context.getDrawable(drawable.ic_btc_icon)
    }

    override fun toFiat(conversionFiat: FiatCurrency): FiatCurrency {
        return toUSD(conversionFiat)
    }


    companion object {
        const val SUB_NUM_MAX: Int = 8
        const val ALT_CURRENCY_FORMAT = "#,##0.00000000 BTC"
        const val symbol = "\u20BF"
        const val maxLongValue = 2099999997690000L

        @JvmField
        val CREATOR: Creator<BTCCurrency?> = object : Creator<BTCCurrency?> {
            override fun createFromParcel(parcel: Parcel): BTCCurrency {
                return BTCCurrency(parcel)
            }

            override fun newArray(size: Int): Array<BTCCurrency?> {
                return arrayOfNulls<BTCCurrency?>(size)
            }
        }
        private val DEFAULT_CURRENCY_FORMAT = String.format("%s #,##0.########", symbol)
        const val format = "#,##0.########"
    }
}